### ISSUE TITLE
Let bundles provide credential hints

### ISSUE THREAD

**Issue by carolynvs** [deislabs/porter#326]
It would be super useful if a bundle could provide hints about the credentials it wants, such suggested defaults on where to find the values and explanatory text for what the credential is.

```yaml
credentials:
- name: kubeconfig
  path: /root/.kube/config
  default: ~/.kube/config
  description: File with embedded server and user certificates to connect to a kubernetes cluster. Helm should have already been init'd on this cluster.
- name: SUBSCRIPTION_ID
  env: AZURE_SUBSCRIPTION_ID
```

**Comment by carolynvs** [comment]
This involves changes in:

* https://github.com/deislabs/porter/blob/master/pkg/credentialsgenerator/generator.go
* [Manifest.CredentialDefinition](https://github.com/deislabs/porter/blob/b24cdd412147c8156942f6d8accb489e71e48e6f/pkg/config/manifest.go#L107)

**Comment by phillipahereza** [comment]
Hey @carolynvs can I take a shot at this

**Comment by carolynvs** [comment]
Sure! I'll assign it to you

**Comment by phillipahereza** [comment]
@carolynvs can I use https://github.com/manifoldco/promptui to implement the `use arrows to move` and `type to filter` functionality?

**Comment by carolynvs** [comment]
@phillipahereza Oops sorry for the slow reply. The command is already implemented, so we have a library in place to manage that feature that you are asking about (use arrows to type and type to filter). I believe we use `gopkg.in/AlecAivazis/survey.v1`.

For this feature, we need to add the credentials description to the existing credentials prompt / interactive flow, which is already stored in the manifest. In addition I'd like to add a default to credentials in the manifest, and then use that to default the answer in the prompt.

I think that the existing library can support this. If you run into trouble with that library, let us know and we can figure out a different library.

**Comment by phillipahereza** [comment]
@carolynvs why are you using `bundle.Credential` instead of `Manifest.CredentialDefinition` in the `GenerateOptions` struct? Why is `Manifest.CredentialDefinition` added to Porter yet it's exactly the same as `build.Credential` and still not being used

**Comment by carolynvs** [comment]
For some backstory on why porter's code looks the way it does:

Porter maintains its own set of structs for the porter manifest. In some cases those structs have the same set of fields as the structs in https://github.com/deislabs/cnab-go (initially they didn't but over time they may have converged because porter existed before cnab-go).

As the CNAB specification has reached 1.0 recently and the implementation in cnab-go has stabilized, we have begun moving some of our code upstream into cnab-go. Or in some cases adding yaml tags to the structs in cnab-go so that we can use it in our manifest.

It has helped Porter move quickly, and sometimes make decisions on how we want to model our manifest in ways that aren't a 1:1 direct mapping to the bundle model that is represented in cnab-go to keep our own representation at times, so we are not in a rush to get rid of our own structs.

---

Specifically for the `porter generate credentials` command, I'm going to defer to @jeremyrickard since he implemented that one. I am guessing he picked the structs that he did because it needs to work for both generic CNAB and Porter bundles, but I couldn't say more without researching the code.

**Comment by phillipahereza** [comment]
So adding the `Default` field to `Manifest.CredentialDefintion` means we'll need to make changes in quite a number of places in the code but on the other hand, if we add `Default` to https://github.com/deislabs/cnab-go, and omit if empty, we only need to change the generator function. Also, we could maybe get rid of all the duplicated structs between the 2 codebases

**Comment by phillipahereza** [comment]
It makes more sense to move the `Default` field to `bundle.Credential` and add the necessary yaml tags to omit if empty. The changes in Porter would then be in the generator function

**Comment by carolynvs** [comment]
If you would like to make changes to the structs in cnab-go, that would also require corresponding changes to the CNAB spec first before the code in cnab-go can be changed.

If you'd like to try to propose a change to the spec, here's how to go about it:
1. Open an issue in https://github.com/deislabs/cnab-spec. Explain the change that you would like to make and why. The reason can't just be to help porter though. So you would need to explain the underlying use case and how it helps CNAB tools in general generate credentials for people who may not always know what are good defaults or where to get these values. For example someone who is new to kubernetes or a cloud provider following a quick start and accepting suggested defaults, such as storing their kubeconfig in ~/.kube/config or their azure credentials in the standard environment variables for azure.
2. If people seem receptive, open a pull request and edit the corresponding files in the cnab-spec repository. Here is a representative pull request that shows what files are changed when a field is added/removed from the core spec for bundles: https://github.com/deislabs/cnab-spec/pull/233/files
3. Wait for 2 core spec maintainers to approve. If you get stuck, you can ask for reviews in the CNCF #cnab channel or ask one of us to help nudge people.
4. Once the change is merged into the cnab-spec repo, then you can submit a pull request to cnab-go making the change, along with unit tests to make sure the field is being marshaled and unmarshaled properly.
5. After your PR to cnab-go is merged, ask someone (the porter maintainers also happen to be cnab-go maintainers) to tag a release of cnab-go.
6. Now in porter you can bump the cnab-go dependency and use it in porter.

The entire process can take a bit of time, just to set expectations. But I don't want to discourage you from participating since it's pretty cool to put new stuff into the spec! 💪 

What I personally do is make the change in Porter first, ensure that I like it and don't need any other changes to support the feature I need (probably not necessary in this case, it's a straightforward feature), and then suggest the feature upstream to the spec. That way I don't request a change, then realize I needed it to be slightly different 2 months later. 😀 

If you would like to try this out, I suggest making the change to the cnab-go repo first locally, make sure that you can get it to work as you just laid out to me in your comment. If you aren't sure how to try out your changes you can: make a fork of cnab-go, push your changes to a branch and update Gopkg.toml to point to your branch and run dep ensure. Then once you know it will work, then start the process with the cnab-spec issue.

**Comment by carolynvs** [comment]
We can make this change first, vet that it's useful, and THEN submit a PR to the CNAB spec for approval/adoption later.

We just realized that we already have description so we just need to add defaultSource which is the default location where a credential is often found, such as an environment variable name, or file path. So what someone would need to do to implement this feature is:

1. Add a defaultSource field to the CredentialDefinition struct in the porter.yaml file in pkg/manifest/manifest.go. We use default for parameters but that is the default _value_ and this is for a source location to find a value, they aren't the same thing.
2. Update the credential set generator to use the defaultSource when the user doesn't pick one. https://github.com/deislabs/porter/blob/master/pkg/credentialsgenerator/generator.go. The prompt/query UI should display the defaultSource so the user knows what will be used if they do not change the default selection in the UI.
3. Update the porter.yaml to bundle.json converter so that it includes the description and defaultSource field in https://github.com/getporter/porter/blob/release/v1/pkg/cnab/config-adapter/adapter.go and update the testdata and cases in the corresponding adpater_test.go
4. Update the `porter explain` command in pkg/porter/explain.go so that when it is run with --output json or yaml, that the defaultSource is included for credentials.
